readium/navigator/src/main/java/org/readium/r2/navigator/media2/MediaSessionNavigator.kt

```diff
@@ -0,0 +1,327 @@
+        Timber.d("Go to locator $locator")
+        val itemIndex = checkNotNull(publication.readingOrder.indexOfFirstWithHref(locator.href))
+        val position = locator.locations.time ?: Duration.ZERO
+        return seek(itemIndex, position)
+    }
+
+    /**
+     * Seeks to the beginning of the given link.
+     */
+    suspend fun go(link: Link) =
+        go(link.toLocator())
+
+    /**
+     * Skips to a little amount of time later.
+     */
+    suspend fun goForward(): Try<Unit, Exception> =
+        seekBy(configuration.skipForwardInterval)
+
+    /**
+     * Skips to a little amount of time before.
+     */
+    suspend fun goBackward(): Try<Unit, Exception> =
+        seekBy(-configuration.skipBackwardInterval)
+
+    private suspend fun seekBy(offset: Duration): Try<Unit, Exception> =
+        this.playerFacade.playlist!!.durations
+            ?.let { smartSeekBy(offset, it) }
+            ?: dummySeekBy(offset)
+
+    private suspend fun smartSeekBy(
+        offset: Duration,
+        durations: List<Duration>
+    ): Try<Unit, Exception> {
+        val (newIndex, newPosition) =
+            SmartSeeker.dispatchSeek(
+                offset,
+                playerFacade.currentPosition!!,
+                playerFacade.currentIndex!!,
+                durations
+            )
+        Timber.d("Smart seeking by $offset resolved to item $newIndex position $newPosition")
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    private suspend fun dummySeekBy(offset: Duration): Try<Unit, Exception> {
+        val newIndex = playerFacade.currentIndex!!
+        val newPosition = playerFacade.currentPosition!! + offset
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    /**
+     * Stops the playback.
+     *
+     * Compared to [pause], the navigator may clear its state in whatever way is appropriate. For
+     * example, recovering a player's resources.
+     */
+    fun close() {
+        playerFacade.unregisterPlayerCallback(playerCallback)
+        playerCallback.close()
+        playerFacade.close()
+        coroutineScope.cancel()
+    }
+
+    fun session(context: Context, id: String, activityIntent: PendingIntent): MediaSession =
+        playerFacade.session(context, id, activityIntent)
+
+
+    data class Configuration(
+        val positionRefreshRate: Double = 2.0,  // Hz
+        val skipForwardInterval: Duration = 30.seconds,
+        val skipBackwardInterval: Duration = 30.seconds,
+    )
+
+    @ExperimentalTime
+    data class Playback(
+        val state: State,
+        val rate: Double,
+        val currentIndex: Int,
+        val currentLink: Link,
+        val currentPosition: Duration,
+        val bufferedPosition: Duration,
+    ) {
+
+        enum class State {
+            Playing,
+            Paused,
+            Finished,
+            Error
+        }
+    }
+
+    sealed class Exception : kotlin.Exception()
```

Comment:
The player doesn't provide exceptions, I can't make them up....

Reply:
Yes we don't need one exception per media3 error code, but we could use a `PlayerException` that wraps the raw code, or another one when the user provides an invalid locator to `go()`, or another wrapping `PublicationDataSource.Exception` if we get it at some point.

Regarding the usefulness, as a toolkit we need to provide as much details as possible for the (nested) error causes to help implementers find out the root cause of an issue and choose whether (or not) and how to recover from it, depending on their strategy.

I recommend this great article about the difference between app-level and library-level error handling: https://nick.groenen.me/posts/rust-error-handling/#libraries-versus-applications

> It took me a moment to appreciate this distinction, but there’s value in approaching error handling differently between libraries and applications as they tend to have different concerns:
> * Libraries should focus on producing meaningful, structured error types/variants. This allows applications to easily differentiate various error cases.
> * Applications mainly consume errors.
> * Libraries may want to cast errors from one type to another. An IO error should likely be wrapped by a high-level error type provided by the library.
>      * Otherwise an IO error in library `foo` cannot be distinguished from a similar IO error in library `bar`.
> * Where libraries return errors, applications decide if and how those errors are formatted and displayed to users.
> * Applications may also want to parse and inspect errors, for example to forward them to exception tracking services or to retry operations when doing so is deemed to be safe.

Reply:
As a fan of Rust, I knew this article! It reminds me that I'm not sure about  the attempts to provide error messages in the toolkit (`UserException`).

About `PublicationDataSource.Exception`, I feel like the `SessionPlayer` would go into the error state if any occur, and that's it. Let me know if you have more insights.

Reply:
> It reminds me that I'm not sure about the attempts to provide error messages in the toolkit (`UserException`).

That's fair, I started this for the LCP errors which can be quite complicated and require some specific mentions in the user messages to pass the certification.

But I agree that returning `UserException` should not be required (or maybe even recommended) for other APIs. As long as we have properly typed and detailed ones and don't return `kotlin.Exception`.